### PR TITLE
feat(protocol-designer, components): finish functionality for selecting tips

### PR DESCRIPTION
**Pull request by ncdiehl11**
## Overview

This PR completes the implementation of tip selection in Protocol Designer (PD) by wiring up UI so users can click, hover, and otherwise interact with tip positions. It updates both the labware components and the tip selection wizard flow to reflect selected, used, and error tip states. It also adds new types to support the new tip selection behavior. Last, it implements safety checks for physically available tip pickups (considering both the current tiprack's tip presence, and the surrounding slots' respective stack heights).

NOTE: Single channel overlays do not currently show. This will be fixed once https://github.com/Opentrons/opentrons/pull/19814 is merged

[Cloudfront test link](https://d2am11gct1emmn.cloudfront.net/pd_feat-select-tips-compatibility/)

## Changelog

- LabwareRender now accepts selectedTipsByIndex and renders clickable wells accordingly. 
- StaticLabware and related internal components now pass through tip selection props (e.g. selectedTipsByIndex) so that selection state is visualized. 
- SelectedTip component’s logic updated to support error and used states (e.g. stroke logic) 
- TipStatus supports new statuses: SELECTED_USED, SELECTED_ERROR, etc. 
- ListButton component now handles an “error” type for tip list UI feedback. 
- Shadows for pipette representations (e.g. eight-channel, ninety-six, flex) updated to reflect tip selection states (some deletions / cleanups in shadow components) 
- general refactors and code enhancements

## Test Plan

Given the size of this PR, I recommend generally smoke testing with various combinations of pipettes and nozzle configurations. 

- In the Protocol Designer UI, open a moveLiquid or mix step.
- Select a nozzle configuration if an 8- or 96-channel pipette.
- Hover wells on the tiprack and observe the correct pipette overlay state based on the presence and accessibility of the tip.
- Click wells and verify that they are selected, and the pickup counter goes down.
- Try to save with fewer than the number of required pickups, and verify that the error banner appears
- Verify un-select behavior by clicking on a selected well (NOTE: all proceeding selections will also be unselected)
- Test edge cases: selecting already used tips, selecting invalid wells.
- Observe that localization strings show the correct pluralization and messages (e.g. “2 pickups selected”, “Used tip”, etc.)

Regression checks
- building liquid-handling steps with "Automatic" tip tracking should behave the same as before

If the team has unit or component tests in PD, tests should ideally be added to cover tip selection logic (e.g. for SelectedTip, TipStatus, and interaction hooks).

## Review requests

see test plan. please let me know if you'd like to walk through this live, since it's quite a large PR

## Risk assessment

medium. core logic touched, but most additions are in new components safely hidden behind FF

### CODE REVIEW

**Comment by jerader** [comment]
starting to smoke test right now! just noticed something: if you select "automatic" tip tracking, it skips over the tip that has been used before, that's expected?

**Comment by jerader** [comment]
found a bug that is somewhat unrelated but leads to an issue with the tip selection wizard. The nozzles defaults to "All" even when you don't have a tiprack you can use. Then when you get to the tip selection wizard, it doesn't let you select a tiprack and doesn't indicate that there is no tiprack available.


https://github.com/user-attachments/assets/804035c8-ca59-4ace-a9e8-a0fa0e6de846

File: protocol-designer/src/assets/localization/en/tip_selection.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "click_and_drag": "Click and drag to select tips in {{labwareName}}",
```
Comment (jerader) [inline protocol-designer/src/assets/localization/en/tip_selection.json:2]: ah we are punting on the drag feature for now?

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/TipTrackingField.tsx [hunk 2]
```diff
@@ -9,25 +10,145 @@ import {
...
+  let airGapByVolume: Array<[number, number]> = []
+  // no air gap included for mix step
+  if (formData.stepType === 'moveLiquid') {
+    airGapByVolume =
+      (liquidClassValuesForTip?.aspirate.retract.airGapByVolume as Array<
+        [number, number]
+      >) ?? []
+  }
+
+  const transferPlanAndReferenceVolumes =
```
Comment (jerader) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/TipTrackingField.tsx:83]: nit: would love to see this logic in a util along with a unit test!

File: step-generation/src/utils/safePipetteMovements.ts [hunk 3]
```diff
@@ -352,38 +362,139 @@ export const getIsSafePipetteMovement = (args: {
...
-        pipetteBoundsAtWellLocation,
-        slotInfos,
-        robotState,
-        invariantContext,
-        labwareId
-      )
+  )
+}
+
+export const getIsSafePickupWithinTiprack = (args: {
```
Comment (jerader) [inline step-generation/src/utils/safePipetteMovements.ts:398]: nit: would be nice to have some unit testing for this util as well

**Review by jerader — COMMENTED** [review]
left some questions but overall this is incredible! the shadow for tips unavailable with the partial tip pick up is really cool. Amazing work!!!

File: protocol-designer/src/assets/localization/en/tip_selection.json (hunk 1 above)
Comment (ncdiehl11) [reply protocol-designer/src/assets/localization/en/tip_selection.json:2]: Click and drag will only be enabled in certain circumstances. I will add logic in a followup to determine if click and drag is allowable and and the feature

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/TipTrackingField.tsx (hunk 2 above)
Comment (ncdiehl11) [reply protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/TipTrackingField.tsx:83]: good call, refactoring the following logic to a util and will add a unit test

**Comment by ncdiehl11** [comment]
> starting to smoke test right now! just noticed something: if you select "automatic" tip tracking, it skips over the tip that has been used before, that's expected?

Nothing is wired up in step generation yet, so yes! I think that will be the expected behavior anyway though

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/TipTrackingField.tsx [hunk 4]
```diff
@@ -9,25 +10,112 @@ import {
...
+  const tiprackEntities = useSelector(getLabwareEntities)
+  const pipette = pipetteEntities[formData.pipette]
+  const tiprackDefinition = Object.values(tiprackEntities).find(
+    tiprackEntity => tiprackEntity.labwareDefURI === formData.tipRack
+  )?.def
+
+  const allLiquidClassDefs = getAllLiquidClassDefs()
+  const liquidClassDef =
+    allLiquidClassDefs[formData.liquidClass ?? ''] ??
+    allLiquidClassDefs[WATER_LIQUID_CLASS_NAME]
```
Comment (jerader) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/TipTrackingField.tsx:64]: seems reasonable to use water liquid class by default

**Review by jerader — APPROVED** [review]
amazing. spectacular. sublime.
